File: sample_apps/js/crud-and-simple-ingestion-example.js
import { constants } from "./constants.js";
import {
    CreateDatabaseCommand, CreateTableCommand, DeleteDatabaseCommand, DeleteTableCommand,
    DescribeDatabaseCommand, DescribeTableCommand,
    ListDatabasesCommand, ListTablesCommand,
    UpdateDatabaseCommand, UpdateTableCommand, WriteRecordsCommand
} from "@aws-sdk/client-timestream-write";

export async function createDatabase(writeClient) {
    console.log("Creating Database");
    const params = new CreateDatabaseCommand({
        DatabaseName: constants.DATABASE_NAME
    });

    await writeClient.send(params).then(
        (data) => {
            console.log(`Database ${data.Database.DatabaseName} created successfully`);
        },
        (err) => {
            if (err.code === 'ConflictException') {
                console.log(`Database ${params.DatabaseName} already exists. Skipping creation.`);
            } else {
                console.log("Error creating database", err);
            }
        }
    );
}

export async function describeDatabase (writeClient) {
    console.log("Describing Database");
    const params = new DescribeDatabaseCommand({
        DatabaseName: constants.DATABASE_NAME
    });

    await writeClient.send(params).then(
        (data) => {
            console.log(`Database ${data.Database.DatabaseName} has id ${data.Database.Arn}`);
        },
        (err) => {
            if (err.code === 'ResourceNotFoundException') {
                console.log("Database doesn't exists.");
            } else {
                console.log("Describe database failed.", err);
                throw err;
            }
        }
    );
}

export async function updateDatabase(updatedKmsKeyId) {

    if (updatedKmsKeyId === undefined) {
        console.log("Skipping UpdateDatabase; KmsKeyId was not given");
        return;
    }
    console.log("Updating Database");
    const params = new UpdateDatabaseCommand({
        DatabaseName: constants.DATABASE_NAME,
        KmsKeyId: updatedKmsKeyId
    });

    await writeClient.send(params).then(
        (data) => {
            console.log(`Database ${data.Database.DatabaseName} updated kmsKeyId to ${updatedKmsKeyId}`);
        },
        (err) => {
            if (err.code === 'ResourceNotFoundException') {
                console.log("Database doesn't exist.");
            } else {
                console.log("Update database failed.", err);
            }
        }
    );
}

export async function listDatabases(writeClient) {
    console.log("Listing databases:");
    const databases = await getDatabasesList(null, writeClient);
    databases.forEach(function(database){
        console.log(database.DatabaseName);
    });
}

function getDatabasesList(nextToken, writeClient, databases = []) {
    var params = new ListDatabasesCommand({
        MaxResults: 15
    });

    if(nextToken) {
        params.NextToken = nextToken;
    }

    return writeClient.send(params).then(
            (data) => {
                databases.push.apply(databases, data.Databases);
                if (data.NextToken) {
                    return getDatabasesList(data.NextToken, databases);
                } else {
                    return databases;
                }
            },
            (err) => {
                console.log("Error while listing databases", err);
            });
}

export async function createTable(writeClient) {
    console.log("Creating Table");
    const params = new CreateTableCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        RetentionProperties: {
            MemoryStoreRetentionPeriodInHours: constants.HT_TTL_HOURS,
            MagneticStoreRetentionPeriodInDays: constants.CT_TTL_DAYS
        }
    });

    await writeClient.send(params).then(
        (data) => {
            console.log(`Table ${data.Table.TableName} created successfully`);
        },
        (err) => {
            if (err.code === 'ConflictException') {
                console.log(`Table ${params.TableName} already exists on db ${params.DatabaseName}. Skipping creation.`);
            } else {
                console.log("Error creating table. ", err);
                throw err;
            }
        }
    );
}

export async function updateTable(writeClient) {
    console.log("Updating Table");
    const params = new UpdateTableCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        RetentionProperties: {
            MemoryStoreRetentionPeriodInHours: constants.HT_TTL_HOURS,
            MagneticStoreRetentionPeriodInDays: constants.CT_TTL_DAYS
        }
    });

    await writeClient.send(params).then(
        (data) => {
            console.log("Table updated")
        },
        (err) => {
            console.log("Error updating table. ", err);
            throw err;
        }
    );
}

export async function describeTable(writeClient) {
    console.log("Describing Table");
    const params = new DescribeTableCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME
    });

    await writeClient.send(params).then(
        (data) => {
            console.log(`Table ${data.Table.TableName} has id ${data.Table.Arn}`);
        },
        (err) => {
            if (err.code === 'ResourceNotFoundException') {
                console.log("Table or Database doesn't exists.");
            } else {
                console.log("Describe table failed.", err);
                throw err;
            }
        }
    );
}

export async function listTables(writeClient) {
    console.log("Listing tables:");
    const tables = await getTablesList(null, writeClient);
    tables.forEach(function(table){
        console.log(table.TableName);
    });
}

function getTablesList(nextToken, writeClient, tables = []) {
    var params = new ListTablesCommand({
        DatabaseName: constants.DATABASE_NAME,
        MaxResults: 15
    });

    if(nextToken) {
        params.NextToken = nextToken;
    }

    return writeClient.send(params).then(
            (data) => {
                tables.push.apply(tables, data.Tables);
                if (data.NextToken) {
                    return getTablesList(data.NextToken, tables);
                } else {
                    return tables;
                }
            },
            (err) => {
                console.log("Error while listing databases", err);
            });
}

export async function writeRecords(writeClient) {
    console.log("Writing records");
    const currentTime = Date.now().toString(); // Unix time in milliseconds

    const dimensions = [
        {'Name': 'region', 'Value': 'us-east-1'},
        {'Name': 'az', 'Value': 'az1'},
        {'Name': 'hostname', 'Value': 'host1'}
    ];

    const cpuUtilization = {
        'Dimensions': dimensions,
        'MeasureName': 'cpu_utilization',
        'MeasureValue': '13.5',
        'MeasureValueType': 'DOUBLE',
        'Time': currentTime.toString()
    };

    const memoryUtilization = {
        'Dimensions': dimensions,
        'MeasureName': 'memory_utilization',
        'MeasureValue': '40',
        'MeasureValueType': 'DOUBLE',
        'Time': currentTime.toString()
    };

    const records = [cpuUtilization, memoryUtilization];

    const params = new WriteRecordsCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        Records: records
    });

    await writeClient.send(params).then(
        (data) => {
            console.log("Write records successful");
        },
        (err) => {
            console.log("Error writing records:", err);
            if (err.code === 'RejectedRecordsException') {
                printRejectedRecordsException(request);
            }
        }
    );
}

export async function writeRecordsWithCommonAttributes(writeClient) {
    console.log("Writing records with common attributes");
    const currentTime = Date.now().toString(); // Unix time in milliseconds

    const dimensions = [
        {'Name': 'region', 'Value': 'us-east-1'},
        {'Name': 'az', 'Value': 'az1'},
        {'Name': 'hostname', 'Value': 'host1'}
    ];

    const commonAttributes = {
        'Dimensions': dimensions,
        'MeasureValueType': 'DOUBLE',
        'Time': currentTime.toString()
    };

    const cpuUtilization = {
        'MeasureName': 'cpu_utilization',
        'MeasureValue': '13.5'
    };

    const memoryUtilization = {
        'MeasureName': 'memory_utilization',
        'MeasureValue': '40'
    };

    const records = [cpuUtilization, memoryUtilization];

    const params = new WriteRecordsCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        Records: records,
        CommonAttributes: commonAttributes
    });

    await writeClient.send(params).then(
        (data) => {
            console.log("Write records successful");
        },
        (err) => {
            console.log("Error writing records:", err);
            if (err.code === 'RejectedRecordsException') {
                printRejectedRecordsException(request);
            }
        }
    );
}

export async function writeRecordsWithUpsert(writeClient) {
    console.log("Writing records with upsert");
    const currentTime = Date.now().toString(); // Unix time in milliseconds
    // To achieve upsert (last writer wins) semantic, one example is to use current time as the version if you are writing directly from the data source
    let version = Date.now();

    const dimensions = [
        {'Name': 'region', 'Value': 'us-east-1'},
        {'Name': 'az', 'Value': 'az1'},
        {'Name': 'hostname', 'Value': 'host1'}
    ];

    const commonAttributes = {
        'Dimensions': dimensions,
        'MeasureValueType': 'DOUBLE',
        'Time': currentTime.toString(),
        'Version': version
    };

    const cpuUtilization = {
        'MeasureName': 'cpu_utilization',
        'MeasureValue': '13.5'
    };

    const memoryUtilization = {
        'MeasureName': 'memory_utilization',
        'MeasureValue': '40'
    };

    const records = [cpuUtilization, memoryUtilization];

    const params = new WriteRecordsCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        Records: records,
        CommonAttributes: commonAttributes
    });

    // write records for first time
    await writeClient.send(params).then(
        (data) => {
            console.log("Write records successful for first time.");
        },
        (err) => {
            console.log("Error writing records:", err);
            if (err.code === 'RejectedRecordsException') {
                printRejectedRecordsException(request);
            }
        }
    );

    // Successfully retry same writeRecordsRequest with same records and versions, because writeRecords API is idempotent.
    await writeClient.send(params).then(
        (data) => {
            console.log("Write records successful for retry.");
        },
        (err) => {
            console.log("Error writing records:", err);
            if (err.code === 'RejectedRecordsException') {
                printRejectedRecordsException(request);
            }
        }
    );

    // upsert with lower version, this would fail because a higher version is required to update the measure value.
    version--;

    const commonAttributesWithLowerVersion = {
        'Dimensions': dimensions,
        'MeasureValueType': 'DOUBLE',
        'Time': currentTime.toString(),
        'Version': version
    };

    const updatedCpuUtilization = {
        'MeasureName': 'cpu_utilization',
        'MeasureValue': '14.5'
    };

    const updatedMemoryUtilization = {
        'MeasureName': 'memory_utilization',
        'MeasureValue': '50'
    };

    const upsertedRecords = [updatedCpuUtilization, updatedMemoryUtilization];

    const upsertedParamsWithLowerVersion = new WriteRecordsCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        Records: upsertedRecords,
        CommonAttributes: commonAttributesWithLowerVersion
    });

    await writeClient.send(upsertedParamsWithLowerVersion).then(
        (data) => {
            console.log("Write records for upsert with lower version successful");
        },
        (err) => {
            console.log("Error writing records:", err);
            if (err.code === 'RejectedRecordsException') {
                printRejectedRecordsException(upsertRequestWithLowerVersion);
            }
        }
    );
    
    // upsert with higher version as new data in generated
    version = Date.now();

    const commonAttributesWithHigherVersion = {
        'Dimensions': dimensions,
        'MeasureValueType': 'DOUBLE',
        'Time': currentTime.toString(),
        'Version': version
    };

    const upsertedParamsWithHigherVerion = new WriteRecordsCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        Records: upsertedRecords,
        CommonAttributes: commonAttributesWithHigherVersion
    });

    await writeClient.send(upsertedParamsWithHigherVerion).then(
        (data) => {
            console.log("Write records upsert successful with higher version");
        },
        (err) => {
            console.log("Error writing records:", err);
            if (err.code === 'RejectedRecordsException') {
                printRejectedRecordsException(upsertedParamsWithHigherVerion);
            }
        }
    );

}

export async function deleteDatabase(writeClient) {
    console.log("Deleting Database");
    const params = new DeleteDatabaseCommand({
        DatabaseName: constants.DATABASE_NAME
    })

    await writeClient.send(params).then(
        function (data) {
            console.log("Deleted database");
         },
        function(err) {
            if (err.code === 'ResourceNotFoundException') {
                console.log(`Database ${params.DatabaseName} doesn't exists.`);
            } else {
                console.log("Delete database failed.", err);
                throw err;
            }
        }
    );
}

export async function deleteTable(writeClient) {
    console.log("Deleting Table");
    const params = new DeleteTableCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME
    });

    await writeClient.send(params).then(
        function (data) {
            console.log("Deleted table");
        },
        function(err) {
            if (err.code === 'ResourceNotFoundException') {
                console.log(`Table ${params.TableName} or Database ${params.DatabaseName} doesn't exists.`);
            } else {
                console.log("Delete table failed.", err);
                throw err;
            }
        }
    );
}

function printRejectedRecordsException(request) {
    const responsePayload = JSON.parse(request.response.httpResponse.body.toString());
                console.log("RejectedRecords: ", responsePayload.RejectedRecords);
}

File: sample_apps/dotnet/CrudAndSimpleIngestionExample.cs
﻿using System;
using System.Threading.Tasks;
using System.Collections.Generic;
using Amazon.TimestreamWrite;
using Amazon.TimestreamWrite.Model;

namespace TimestreamDotNetSample
{
    public class CrudAndSimpleIngestionExample
    {

        private readonly AmazonTimestreamWriteClient writeClient;

        public CrudAndSimpleIngestionExample(AmazonTimestreamWriteClient writeClient)
        {
            this.writeClient = writeClient;
        }

        public async Task CreateDatabase(string databaseName)
        {
            Console.WriteLine("Creating Database");

            try
            {
                var createDatabaseRequest = new CreateDatabaseRequest
                {
                    DatabaseName = databaseName
                };
                CreateDatabaseResponse response = await writeClient.CreateDatabaseAsync(createDatabaseRequest);
                Console.WriteLine($"Database {databaseName} created");
            }
            catch (ConflictException)
            {
                Console.WriteLine("Database already exists.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Create database failed:" + e.ToString());
            }

        }

        public async Task DescribeDatabase(string databaseName)
        {
            Console.WriteLine("Describing Database");

            try
            {
                var describeDatabaseRequest = new DescribeDatabaseRequest
                {
                    DatabaseName = databaseName
                };
                DescribeDatabaseResponse response = await writeClient.DescribeDatabaseAsync(describeDatabaseRequest);
                Console.WriteLine($"Database {databaseName} has id:{response.Database.Arn}");
            }
            catch (ResourceNotFoundException)
            {
                Console.WriteLine("Database does not exist.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Describe database failed:" + e.ToString());
            }

        }

        public async Task ListDatabases()
        {
            Console.WriteLine("Listing Databases");

            try
            {
                var listDatabasesRequest = new ListDatabasesRequest
                {
                    MaxResults = 5
                };
                ListDatabasesResponse response = await writeClient.ListDatabasesAsync(listDatabasesRequest);
                PrintDatabases(response.Databases);
                var nextToken = response.NextToken;
                while (nextToken != null)
                {
                    listDatabasesRequest.NextToken = nextToken;
                    response = await writeClient.ListDatabasesAsync(listDatabasesRequest);
                    PrintDatabases(response.Databases);
                    nextToken = response.NextToken;
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("List database failed:" + e.ToString());
            }

        }

        public async Task UpdateDatabase(String databaseName, String updatedKmsKeyId)
        {
            Console.WriteLine("Updating Database");

            try
            {
                var updateDatabaseRequest = new UpdateDatabaseRequest
                {
                    DatabaseName = databaseName,
                    KmsKeyId = updatedKmsKeyId
                };
                UpdateDatabaseResponse response = await writeClient.UpdateDatabaseAsync(updateDatabaseRequest);
                Console.WriteLine($"Database {databaseName} updated with KmsKeyId {updatedKmsKeyId}");
            }
            catch (ResourceNotFoundException)
            {
                Console.WriteLine("Database does not exist.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Update database failed: " + e.ToString());
            }

        }

        private void PrintDatabases(List<Database> databases)
        {
            foreach (Database database in databases)
                Console.WriteLine($"Database:{database.DatabaseName}");
        }

        public async Task DeleteDatabase(String databaseName)
        {
            Console.WriteLine("Deleting database");
            try
            {
                var deleteDatabaseRequest = new DeleteDatabaseRequest
                {
                    DatabaseName = databaseName
                };
                DeleteDatabaseResponse response = await writeClient.DeleteDatabaseAsync(deleteDatabaseRequest);
                Console.WriteLine($"Database {databaseName} delete request status:{response.HttpStatusCode}");
            }
            catch (ResourceNotFoundException)
            {
                Console.WriteLine($"Database {databaseName} does not exists");
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception while deleting database:" + e.ToString());
            }
        }

        public async Task CreateTable(String databaseName, String tableName)
        {
            Console.WriteLine("Creating Table");

            try
            {
                var createTableRequest = new CreateTableRequest
                {
                    DatabaseName = databaseName,
                    TableName = tableName,
                    RetentionProperties = new RetentionProperties
                    {
                        MagneticStoreRetentionPeriodInDays = Constants.CT_TTL_DAYS,
                        MemoryStoreRetentionPeriodInHours = Constants.HT_TTL_HOURS
                    }
                };
                CreateTableResponse response = await writeClient.CreateTableAsync(createTableRequest);
                Console.WriteLine($"Table {tableName} created");
            }
            catch (ConflictException)
            {
                Console.WriteLine("Table already exists.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Create table failed:" + e.ToString());
            }

        }

        public async Task DescribeTable(String databaseName, String tableName)
        {
            Console.WriteLine("Describing Table");

            try
            {
                var describeTableRequest = new DescribeTableRequest
                {
                    DatabaseName = databaseName,
                    TableName = tableName
                };
                DescribeTableResponse response = await writeClient.DescribeTableAsync(describeTableRequest);
                Console.WriteLine($"Table {tableName} has id:{response.Table.Arn}");
            }
            catch (ResourceNotFoundException)
            {
                Console.WriteLine("Table does not exist.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Describe table failed:" + e.ToString());
            }

        }

        public async Task ListTables(String databaseName)
        {
            Console.WriteLine("Listing Tables");

            try
            {
                var listTablesRequest = new ListTablesRequest
                {
                    MaxResults = 5,
                    DatabaseName = databaseName
                };
                ListTablesResponse response = await writeClient.ListTablesAsync(listTablesRequest);
                PrintTables(response.Tables);
                string nextToken = response.NextToken;
                while (nextToken != null)
                {
                    listTablesRequest.NextToken = nextToken;
                    response = await writeClient.ListTablesAsync(listTablesRequest);
                    PrintTables(response.Tables);
                    nextToken = response.NextToken;
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("List table failed:" + e.ToString());
            }

        }

        private void PrintTables(List<Table> tables)
        {
            foreach (Table table in tables)
                Console.WriteLine($"Table: {table.TableName}");
        }

        public async Task UpdateTable(String databaseName, String tableName)
        {
            Console.WriteLine("Updating Table");

            try
            {
                var updateTableRequest = new UpdateTableRequest
                {
                    DatabaseName = databaseName,
                    TableName = tableName,
                    RetentionProperties = new RetentionProperties
                    {
                        MagneticStoreRetentionPeriodInDays = Constants.CT_TTL_DAYS,
                        MemoryStoreRetentionPeriodInHours = Constants.HT_TTL_HOURS
                    }
                };
                UpdateTableResponse response = await writeClient.UpdateTableAsync(updateTableRequest);
                Console.WriteLine($"Table {tableName} updated");
            }
            catch (ResourceNotFoundException)
            {
                Console.WriteLine("Table does not exist.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Update table failed:" + e.ToString());
            }

        }

        public async Task DeleteTable(String databaseName, String tableName)
        {
            Console.WriteLine("Deleting table");
            try
            {
                var deleteTableRequest = new DeleteTableRequest
                {
                    DatabaseName = databaseName,
                    TableName = tableName
                };
                DeleteTableResponse response = await writeClient.DeleteTableAsync(deleteTableRequest);
                Console.WriteLine($"Table {tableName} delete request status: {response.HttpStatusCode}");
            }
            catch (ResourceNotFoundException)
            {
                Console.WriteLine($"Table {tableName} does not exists");
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception while deleting table:" + e.ToString());
            }
        }

        public async Task WriteRecords(String databaseName, String tableName)
        {
            Console.WriteLine("Writing records");

            DateTimeOffset now = DateTimeOffset.UtcNow;
            string currentTimeString = (now.ToUnixTimeMilliseconds()).ToString();

            List<Dimension> dimensions = new List<Dimension>{
                new Dimension { Name = "region", Value = "us-east-1" },
                new Dimension { Name = "az", Value = "az1" },
                new Dimension { Name = "hostname", Value = "host1" }
            };

            var cpuUtilization = new Record
            {
                Dimensions = dimensions,
                MeasureName = "cpu_utilization",
                MeasureValue = "13.6",
                MeasureValueType = MeasureValueType.DOUBLE,
                Time = currentTimeString
            };

            var memoryUtilization = new Record
            {
                Dimensions = dimensions,
                MeasureName = "memory_utilization",
                MeasureValue = "40",
                MeasureValueType = MeasureValueType.DOUBLE,
                Time = currentTimeString
            };


            List<Record> records = new List<Record> {
                cpuUtilization,
                memoryUtilization
            };

            try
            {
                var writeRecordsRequest = new WriteRecordsRequest
                {
                    DatabaseName = databaseName,
                    TableName = tableName,
                    Records = records
                };
                WriteRecordsResponse response = await writeClient.WriteRecordsAsync(writeRecordsRequest);
                Console.WriteLine($"Write records status code: {response.HttpStatusCode.ToString()}");
            }
            catch (RejectedRecordsException e) {
                PrintRejectedRecordsException(e);
            }
            catch (Exception e)
            {
                Console.WriteLine("Write records failure:" + e.ToString());
            }
        }

        public async Task WriteRecordsWithCommonAttributes(String databaseName, String tableName)
        {
            Console.WriteLine("Writing records with common attributes");

            DateTimeOffset now = DateTimeOffset.UtcNow;
            string currentTimeString = (now.ToUnixTimeMilliseconds()).ToString();

            List<Dimension> dimensions = new List<Dimension>{
                new Dimension { Name = "region", Value = "us-east-1" },
                new Dimension { Name = "az", Value = "az1" },
                new Dimension { Name = "hostname", Value = "host1" }
            };

            var commonAttributes = new Record
            {
                Dimensions = dimensions,
                MeasureValueType = MeasureValueType.DOUBLE,
                Time = currentTimeString
            };

            var cpuUtilization = new Record
            {
                MeasureName = "cpu_utilization",
                MeasureValue = "13.6"
            };

            var memoryUtilization = new Record
            {
                MeasureName = "memory_utilization",
                MeasureValue = "40"
            };


            List<Record> records = new List<Record>();
            records.Add(cpuUtilization);
            records.Add(memoryUtilization);

            try
            {
                var writeRecordsRequest = new WriteRecordsRequest
                {
                    DatabaseName = databaseName,
                    TableName = tableName,
                    Records = records,
                    CommonAttributes = commonAttributes
                };
                WriteRecordsResponse response = await writeClient.WriteRecordsAsync(writeRecordsRequest);
                Console.WriteLine($"Write records status code: {response.HttpStatusCode.ToString()}");
            }
            catch (RejectedRecordsException e) {
                PrintRejectedRecordsException(e);
            }
            catch (Exception e)
            {
                Console.WriteLine("Write records failure:" + e.ToString());
            }
        }

        public async Task WriteRecordsWithUpsert(String databaseName, String tableName)
            {
                Console.WriteLine("Writing records with upsert");

                DateTimeOffset now = DateTimeOffset.UtcNow;
                string currentTimeString = (now.ToUnixTimeMilliseconds()).ToString();
                // To achieve upsert (last writer wins) semantic, one example is to use current time as the version if you are writing directly from the data source
                long version = now.ToUnixTimeMilliseconds();

                List<Dimension> dimensions = new List<Dimension>{
                    new Dimension { Name = "region", Value = "us-east-1" },
                    new Dimension { Name = "az", Value = "az1" },
                    new Dimension { Name = "hostname", Value = "host1" }
                };

                var commonAttributes = new Record
                {
                    Dimensions = dimensions,
                    MeasureValueType = MeasureValueType.DOUBLE,
                    Time = currentTimeString,
                    Version = version
                };

                var cpuUtilization = new Record
                {
                    MeasureName = "cpu_utilization",
                    MeasureValue = "13.6"
                };

                var memoryUtilization = new Record
                {
                    MeasureName = "memory_utilization",
                    MeasureValue = "40"
                };


                List<Record> records = new List<Record>();
                records.Add(cpuUtilization);
                records.Add(memoryUtilization);

                // write records for first time
                try
                {
                    var writeRecordsRequest = new WriteRecordsRequest
                    {
                        DatabaseName = databaseName,
                        TableName = tableName,
                        Records = records,
                        CommonAttributes = commonAttributes
                    };
                    WriteRecordsResponse response = await writeClient.WriteRecordsAsync(writeRecordsRequest);
                    Console.WriteLine($"WriteRecords Status for first time: {response.HttpStatusCode.ToString()}");
                }
                catch (RejectedRecordsException e) {
                    PrintRejectedRecordsException(e);
                }
                catch (Exception e)
                {
                    Console.WriteLine("Write records failure:" + e.ToString());
                }

                // Successfully retry same writeRecordsRequest with same records and versions, because writeRecords API is idempotent.
                try
                {
                    var writeRecordsRequest = new WriteRecordsRequest
                    {
                        DatabaseName = databaseName,
                        TableName = tableName,
                        Records = records,
                        CommonAttributes = commonAttributes
                    };
                    WriteRecordsResponse response = await writeClient.WriteRecordsAsync(writeRecordsRequest);
                    Console.WriteLine($"WriteRecords Status for retry: {response.HttpStatusCode.ToString()}");
                }
                catch (RejectedRecordsException e) {
                    PrintRejectedRecordsException(e);
                }
                catch (Exception e)
                {
                    Console.WriteLine("Write records failure:" + e.ToString());
                }

                // upsert with lower version, this would fail because a higher version is required to update the measure value.
                version--;
                Type recordType = typeof(Record);
                recordType.GetProperty("Version").SetValue(commonAttributes, version);
                recordType.GetProperty("MeasureValue").SetValue(cpuUtilization, "14.6");
                recordType.GetProperty("MeasureValue").SetValue(memoryUtilization, "50");

                List<Record> upsertedRecords = new List<Record> {
                    cpuUtilization,
                    memoryUtilization
                };

                try
                {
                    var writeRecordsUpsertRequest = new WriteRecordsRequest
                    {
                        DatabaseName = databaseName,
                        TableName = tableName,
                        Records = upsertedRecords,
                        CommonAttributes = commonAttributes
                    };
                    WriteRecordsResponse upsertResponse = await writeClient.WriteRecordsAsync(writeRecordsUpsertRequest);
                    Console.WriteLine($"WriteRecords Status for upsert with lower version: {upsertResponse.HttpStatusCode.ToString()}");
                }
                catch (RejectedRecordsException e) {
                    PrintRejectedRecordsException(e);
                }
                catch (Exception e)
                {
                    Console.WriteLine("Write records failure:" + e.ToString());
                }

                // upsert with higher version as new data in generated
                now = DateTimeOffset.UtcNow;
                version = now.ToUnixTimeMilliseconds();
                recordType.GetProperty("Version").SetValue(commonAttributes, version);

                try
                {
                    var writeRecordsUpsertRequest = new WriteRecordsRequest
                    {
                        DatabaseName = databaseName,
                        TableName = tableName,
                        Records = upsertedRecords,
                        CommonAttributes = commonAttributes
                    };
                    WriteRecordsResponse upsertResponse = await writeClient.WriteRecordsAsync(writeRecordsUpsertRequest);
                    Console.WriteLine($"WriteRecords Status for upsert with higher version:  {upsertResponse.HttpStatusCode.ToString()}");
                }
                catch (RejectedRecordsException e) {
                    PrintRejectedRecordsException(e);
                }
                catch (Exception e)
                {
                    Console.WriteLine("Write records failure:" + e.ToString());
                }
            }

        private void PrintRejectedRecordsException(RejectedRecordsException e)
        {
            Console.WriteLine("RejectedRecordsException:" + e.ToString());
            foreach (RejectedRecord rr in e.RejectedRecords) {
                Console.WriteLine("RecordIndex " + rr.RecordIndex + " : " + rr.Reason);
                long? existingVersion = rr.ExistingVersion;
                if (existingVersion != null) {
                    Console.WriteLine("Rejected record existing version: " + existingVersion);
                }
            }
        }
    }
}
